EIPS/eip-7534.md

```diff
@@ -0,0 +1,255 @@
+
+```go
+// Transaction is an Ethereum transaction.
+type Transaction struct {
+	inner TxData    // Consensus contents of a transaction
+	time  time.Time // Time first seen locally (spam avoidance)
+
+	// TCT hypoHash
+	HypoHash common.Hash
+
+	// caches
+	hash atomic.Value
+	size atomic.Value
+	from atomic.Value
+}
+```
+
+Given `HypoHash`, it could retrieve the hypothesis through function `retrieveHypoHash`. And then execute the given transaction in a new EVM environment. The executed evm bytecode will be checked by TCT along with the hypothesis to ensure whether it violates the theorem.
+
+```go
+func applyTransaction(msg *Message, config *params.ChainConfig, gp *GasPool, statedb *state.StateDB, blockNumber *big.Int, blockHash common.Hash, tx *types.Transaction, usedGas *uint64, evm *vm.EVM) (*types.Receipt, error) {
+    ...
+    // TCT: get hypothesis given hypoHash
+    hypothesis, err := retrieveHypoHash(hypoHash)
+    ...
+    // Create a new context to be used in the EVM environment.
+    txContext := NewEVMTxContext(msg)
+    evm.Reset(txContext, statedb)
+
+    // Apply the transaction to the current state (included in the env).
+    result, err := ApplyMessage(evm, msg, gp)
+    ...
+}
+```
+
+## Rationale
+
+Conventional approaches predominantly resort to specialized tools and the expertise of domain professionals to detect and address known vulnerabilities or bugs, targeting the prevention of unintended contract behaviors. While these methods have their merits, empirical evidence suggests a palpable shortfall in ensuring comprehensive safety features. In contrast, our application of TCT brings forth a robust security guarantee via runtime verification, markedly elevating the safety standards compared to the traditional strategies.  
+
+While numerous projects delve into runtime verification, our approach is distinctly tailored to focus solely on transactions which has several merits:
+
+- **Minimal runtime overhead**: By so doing, we anticipate a tangible reduction in the overhead typically associated with blanket verification mechanisms.
+
+- **Concolic testing**: Further, by synthesizing the merits of runtime verification with symbolic proof, our methodology stands out in ensuring both the provability and verifiability of theorems, setting a benchmark in efficacy and reliability. 
+
+
+## Backwards Compatibility
+
+The primary concern lies in the potential for incompatibility, particularly with regards to the occurrence of a fork. The Theorem-Carrying-Transaction (TCT) system may reject certain transactions that violate the theorem, while the original Ethereum network may accept these transactions. This discrepancy in transaction acceptance could lead to a network fork, resulting in divergent block states. To address this issue, we propose conducting a comprehensive evaluation of TCT's effectiveness and stability by initially deploying it on a testnet such as `Goeril`.
```

Comment:
if its a new transaction type enabled at some fork i don't understand how it makes it backward incompatible

Reply:
You are right, it will be incompatible. That's why we want to use a testnet to test TCT first.